src/mo_types/type.ml

```diff
@@ -67,6 +67,205 @@ and kind =
+  | None, Some _ -> -1
+  | Some _, None -> 1
+
+let rec compare_typ (t1 : typ) (t2 : typ) =
+  if t1 == t2 then 0
+  else match (t1, t2) with
+  | Prim p1, Prim p2 ->
+    compare_prim p1 p2
+  | Var (s1, i1), Var (s2, i2) ->
+    (match Int.compare i1 i2 with
+     | 0 -> String.compare s1 s2
+     | ord -> ord)
+  | Con (c1, ts1), Con (c2, ts2) ->
+    (match Cons.compare c1 c2 with
+     | 0 -> compare_typs ts1 ts2
+     | ord -> ord)
+  | Array t1, Array t2 ->
+    compare_typ t1 t2
+  | Tup ts1, Tup ts2 ->
+    compare_typs ts1 ts2
+  | Func (s1, c1, tbs1, ts11, ts12),
+    Func (s2, c2, tbs2, ts21, ts22) ->
+    (match compare_func_sort s1 s2 with
+     | 0 ->
+       (match compare_control c1 c2 with
+        | 0 ->
+          (match compare_tbs tbs1 tbs2 with
+           | 0 ->
+             (match compare_typs ts11 ts21 with
+             | 0 -> compare_typs ts12 ts22
+             | ord -> ord)
+           | ord -> ord)
+        | ord -> ord)
+     | ord -> ord)
+  | Opt t1, Opt t2 -> compare_typ t1 t2
+  | Async (t11, t12) , Async (t21, t22) ->
+    (match compare_typ t11 t21 with
+     | 0 -> compare_typ t12 t22
+     | ord -> ord)
+  | Obj (s1, fs1), Obj (s2, fs2) ->
+    (match compare_obj_sort s1 s2 with
+     | 0 -> compare_flds fs1 fs2
+     | ord -> ord)
+  | Variant fs1, Variant fs2 ->
+    compare_flds fs1 fs2
+  | Mut t1, Mut t2 ->
+    compare_typ t1 t2
+  | Any, Any -> 0
+  | Non, Non -> 0
+  | Pre, Pre -> 0
+  | Typ c1, Typ c2 -> Cons.compare c1 c2
+  | _ -> Int.compare (tag t1) (tag t2)
+
+and compare_tb tb1 tb2 =
+  match String.compare tb1.var tb2.var with
+  | 0 ->
+    (match compare_typ tb1.bound tb2.bound with
+     | 0 -> compare_bind_sort tb1.sort tb2.sort
+     | ord -> ord)
+  | ord ->  ord
+
+and compare_tbs tbs1 tbs2 =
+  match (tbs1, tbs2) with
+  | [], [] -> 0
+  | [], (_::_) -> -1
+  | (_::_, []) -> 1
+  | (tb1::tbs1, tb2 :: tbs2) ->
+    (match compare_tb tb1 tb2 with
+     | 0 -> compare_tbs tbs1 tbs2
+     | ord -> ord)
+
+and compare_fld fld1 fld2 =
+  match String.compare fld1.lab fld2.lab with
+  | 0 ->
+    (match compare_typ fld1.typ fld2.typ with
+     | 0 -> compare_depr fld1.depr fld2.depr
+     | ord -> ord)
+  | ord -> ord
+
+and compare_flds flds1 flds2 =
+  match (flds1, flds2) with
+  | [], [] -> 0
+  | [], (_::_) -> -1
+  | (_::_, []) -> 1
+  | (fld1::flds1, fld2 :: flds2) ->
+    (match compare_fld fld1 fld2 with
+     | 0 -> compare_flds flds1 flds2
+     | ord -> ord)
+
+and compare_typs ts1 ts2 =
+  match (ts1, ts2) with
+  | [], [] -> 0
+  | [], (_::_) -> -1
+  | (_::_, []) -> 1
+  | (t1::ts1, t2 :: ts2) ->
+    (match compare_typ t1 t2 with
+     | 0 -> compare_typs ts1 ts2
+     | ord -> ord)
+
+let compare_rel (t1, t2) (u1, u2) =
+  match compare_typ (t1 : typ) (u1 : typ) with
+  | 0 -> compare_typ (t2 : typ) (u2 : typ)
+  | ord -> ord
+
+(* Syntactic orderings *)
+
+module Ord = struct
+  type t = typ
+  let compare t1 t2 = compare_typ t1 t2
```

Comment:
would eta-reducing speed this up?

Reply:
I doubt it makes much difference since it isn't a recursive entry point, but I can try.

Reply:
Couldn't really tell a difference but shorter code, so better.